outlines/models/llamacpp.py

```diff
@@ -1,304 +1,185 @@
-                add_bos=self.n_past == 0,
-                special=False,
-            )
-
-            self.n_past += n_of_tok
-
-            if n_of_tok > max_len:
-                max_len = n_of_tok
-
-            embd_inp = embd_inp[:n_of_tok]
-            token_ids.append(np.array(embd_inp))
-
-        max_len = np.max([len(a) for a in token_ids])
-        padded = np.asarray(
-            [
-                np.pad(
-                    a,
-                    (0, max_len - len(a)),
-                    "constant",
-                    constant_values=self.pad_token_id,
-                )
-                for a in token_ids
-            ]
-        )
+        for t in range(model.n_vocab()):
+            token_piece = model.tokenizer().decode([t])
+            self.vocabulary[token_piece] = t
+
+    def convert_token_to_string(self, token: str) -> str:
+        return token
 
-        token_ids = torch.LongTensor(padded)
-        return token_ids, torch.ones_like(token_ids)
 
-    def decode(self, token_ids: NDArray[np.int64]) -> List[str]:
-        if isinstance(token_ids, list):
-            token_ids = np.array(token_ids)
-        if token_ids.ndim == 1:
-            token_ids = [token_ids]
+def llamacpp(
+    model_name: str,
+    device: Optional[str] = None,
+    model_kwargs: dict = {},
+):
+    if device == "cuda":
+        model_kwargs["n_gpu_layers"] = -1
 
-        pieces = []
-        for tokens in token_ids:
-            seq = []
-            for id in tokens:
-                size = 32
-                buffer = (ctypes.c_char * size)()
-                n = self.llama_instance.llama_token_to_piece(
-                    self.model, self.llama_instance.llama_token(id), buffer, size
-                )
+    return LlamaCpp(model_name, **model_kwargs)
 
-                token_piece = buffer[:n].decode("utf-8")  # type: ignore
 
-                seq.append(token_piece)
+class LogitsProcessor:
+    def __init__(self, tokenizer: LlamaCppTokenizer, fsm: FSM):
+        """A FSM-based logits processor.
 
-            pieces.append("".join(seq))
+        Parameters
+        ----------
+        tokenizer
+            An instance of `Tokenizer`
+        fsm
+            An instance of `FSM`
 
-        return pieces
+        """
+        self.tokenizer = tokenizer
+        self.fsm_state = FSMState(0)
+        self.fsm: FSM = fsm
+        self.is_first_token = True
 
-    def _get_eos_token(self):
-        size = 32
-        buffer = (ctypes.c_char * size)()
-        n = self.llama_instance.llama_token_to_piece(
-            self.model, self.llama_instance.llama_token(self.eos_token_id), buffer, size
-        )
+    def __call__(
+        self, input_ids: NDArray[np.int64], scores: NDArray[np.float32]
+    ) -> NDArray[np.float32]:
+        """Use the FSM to bias the logits before sampling the next token."""
 
-        token_piece = buffer[:n].decode("utf-8")
+        if self.is_first_token:
+            self.is_first_token = False
```

Comment:
Note: This works so long as we don't expose llama logits processors to the user. If the user uses the same logits processor twice, `self.is_first_token` not being reset will cause problems.

I don't see any docs or other ways in which the logits processor would be exposed, so I think it's fine for now.

Reply:
Would using a token_ids -> state map solve this?

Reply:
I believe so but I haven't tested. This is working for now, so perhaps I should refactor so there is a base logits processor in a separate PR?